Move UseClassicFacingFudge to Mods.Cnc

reaperrr opened the pull request:
Moves the RA/TD-only facing fudge gymnastics out of common mod code to C&C-specific one.

As discussed on IRC.

reaperrr committed:
Move ClassicFacingFudge support to Mods.Cnc

This moves the TD/RA-specific re-mapping of sprite facings
and coordinates to Mods.Cnc.

reaperrr committed:
MoveClassicFacingFudge update rule

pchote commented:
Naming nit: Can we please call these "Classic*" instead of "Legacy*"? The `UseClassicFacingFudge` flag on the body orientation trait can also simplify to `UseFacingFudge` or maybe something like `FudgeFacings`.

pchote commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -368,7 +362,7 @@ public Sprite GetShadow(int frame, int facing)
 		protected virtual Sprite GetSprite(int start, int frame, int facing)
 		{
-			var f = Util.QuantizeFacing(facing, Facings, useClassicFacingFudge);
+			var f = Util.QuantizeFacing(facing, Facings);
```
It would be cleaner if we could define a
```csharp
protected virtual int QuantizeFacing(int facing)
{
	return Util.QuantizeFacing(facing, Facings);
}
```
to call here and override in the subclass. This would save us from duplicating the rest of the calculations below.

pchote commented on OpenRA.Mods.Cnc/Graphics/LegacySpriteSequence.cs:
```diff
@@ -0,0 +1,66 @@
+ */
+#endregion
+
+using System;
```
Only
```csharp
using System;
using OpenRA.Mods.Common.Graphics;
using OpenRA.Graphics;
```
are actually used.

pchote commented on OpenRA.Mods.Cnc/Graphics/LegacyTilesetSpecificSpriteSequence.cs:
```diff
@@ -0,0 +1,93 @@
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Graphics;
```
not used.

pchote commented on OpenRA.Mods.Cnc/LocalUtil.cs:
```diff
@@ -0,0 +1,51 @@
+ */
+#endregion
+
+using System;
```
None(!) of these are used.

pchote commented on OpenRA.Mods.Cnc/Traits/LegacyBodyOrientation.cs:
```diff
@@ -0,0 +1,36 @@
+ */
+#endregion
+
+using System;
```
Only
```csharp
using OpenRA.Mods.Common.Traits;
```
is used.

pchote reviewed (commented):
Some usings cleanup:

reaperrr committed:
Add sequence update rule support

reaperrr committed:
Fixed RA STNK turret not using fudged facings

reaperrr committed:
Misc yaml style fixes

reaperrr commented:
Updated.
In addition to adressing the fixup requests, I also changed the internal default to `true` on `ClassicBodyOrientation.FudgeFacings`, since it doesn't cause any problems with unaffected actors (less than 32 facings) or 32-facing actors in the 2 official mods.
Left the default at `false` for the sequence loaders, though, since you can only use one loader for the entire mod at a time and I don't want to force mods that might mix fudged and non-fudged actors with 32 facings to update all their sequence definitions.

Also added sequence update rule support as part of this.

reaperrr commented:
Updated again.
Diff looks more intimidating than the PR really is, I recommend reviewing commit by commit.

teinarss commented:
Needs a rebase

pchote commented:
We're going to want this and @reaperrr's `facing-1024` branch before we start integrating the ideas from #17796. If you don't have time @reaperrr I can take these over as time allows.

reaperrr commented:
Rebased.

Also updated (but not rebased) the `facing-1024` branch with some additional fixed I'd made locally before losing motivation.

> If you don't have time @reaperrr I can take these over as time allows.

I guess I can manage this one, but it would help if you took over the `facing-1024` branch since you'd likely need far less time to debug and whip it into shape.

Mailaender reviewed (dismissed):
I can't spot regressions in-game and the code looks fine.

pchote commented on OpenRA.Mods.Common/UpdateRules/UpdatePath.cs:
```diff
@@ -74,6 +74,7 @@ public class UpdatePath
 				new RenameSpins(),
 				new CreateScreenShakeWarhead(),
 				new RenameRallyPointPath(),
+				new MoveClassicFacingFudge(),
```
This path now corresponds to the prep-2003 branch - #17808 adds a new "playtest-20200303" path and puts the new rules outside a dated folder until we know the final release date.

We will need to either do the same here, or rebase on top of it after it has been merged.

pchote commented on OpenRA.Mods.Cnc/LocalUtil.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace OpenRA.Mods.Cnc
+{
+	public static class LocalUtil
```
`LocalUtil` isn't very meaningful. Can we call this `FacingUtil` instead?

reaperrr commented:
Rebased and updated.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c132e8b7abad8cc277d683f67cb6312d6f2b68c5...09d59e37a7eef03c208358ee2815f43100f0fc2d)
